californium-core/src/main/java/org/eclipse/californium/core/network/Matcher.java

```diff
@@ -145,33 +159,75 @@ public synchronized void setExecutor(final ScheduledExecutorService executor) {
+
+		// ensure Token is set
+		final KeyToken idByToken = assertOutboundRequestHasToken(exchange, request);
+
+		exchange.setObserver(exchangeObserver);
+		exchangesByMID.put(idByMID, exchange);
+		exchangesByToken.put(idByToken, exchange);
+		addToObservationStore(exchange, request);
+		LOGGER.log(Level.FINE, "tracking open request using [{0}, {1}]", new Object[]{idByMID, idByToken});
+	}
+
+	private KeyMID assertOutboundRequestHasMessageId(final Request request) {
 		if (request.getMID() == Message.NONE) {
-			request.setMID(currendMID.getAndIncrement()%(1<<16)); // wrap at 2^16
+			// request has no message ID set, create a new one
+			request.setMID(currendMID.getAndIncrement() % MAX_MESSAGE_ID); // wrap at 2^16
 		}
-		// request MID is from the local namespace -- use blank address
-		KeyMID idByMID = new KeyMID(request.getMID());
+		return new KeyMID(request.getMID());
+	}
 
-		// ensure Token is set
-		KeyToken idByToken;
+	private KeyToken assertOutboundRequestHasToken(final Exchange exchange, final Request request) {
+		KeyToken result = null;
 		if (request.getToken() == null) {
-			idByToken = createUnusedToken();
-			request.setToken(idByToken.token);
+			result = createUnusedToken();
+			request.setToken(result.token);
 		} else {
-			idByToken = new KeyToken(request.getToken());
+			result = new KeyToken(request.getToken());
+			boolean tokenIsInUseAlready = exchangesByToken.containsKey(result);
+			boolean requestIsPartOfBlockwiseTransfer = request.getOptions().hasBlock1() || request.getOptions().hasBlock2();
+			boolean requestIsRetransmission = exchange.getFailedTransmissionCount() > 0;
 			// ongoing requests may reuse token
-			if (!(exchange.getFailedTransmissionCount()>0 || request.getOptions().hasBlock1() || request.getOptions().hasBlock2() || request.getOptions().hasObserve()) && exchangesByToken.get(idByToken) != null) {
-				LOGGER.log(Level.WARNING, "Manual token overrides existing open request: {0}", idByToken);
+			if (tokenIsInUseAlready && !requestIsRetransmission && !requestIsPartOfBlockwiseTransfer && !request.getOptions().hasObserve()) {
+				LOGGER.log(Level.WARNING, "Manual token overrides existing open request: {0}", result);
 			}
 		}
+		return result;
+	}
 
-		exchange.setObserver(exchangeObserver);
-		LOGGER.log(Level.FINE, "Tracking open request using {0}, {1}", new Object[]{idByMID, idByToken});
+	private void addToObservationStore(final Exchange exchange, final Request request) {
+		boolean isObserveRequest = request.getOptions().hasObserve() && request.getOptions().getObserve() == 0;
+		boolean containsEarlyNegotiationParams = request.getOptions().hasBlock2() && request.getOptions().getBlock2().getNum() == 0 && !request.getOptions().getBlock2().isM();
+		// for observe request.
+		// We ignore blockwise request, except when this is an early negotiation (num and M is set to 0)  
+		if (isObserveRequest || containsEarlyNegotiationParams) {
```

Comment:
Why does it go into the ObservationStore when it containsEarlyNegotiationParams?